src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticAndGeneratorItemSource.cs

```diff
@@ -7,187 +7,208 @@
+    /// <see langword="null"/> again.
+    /// </summary>
+    private AnalyzerReference? _analyzerReference_DoNotAccessDirectly;
+
+    public BaseDiagnosticAndGeneratorItemSource(
+        IThreadingContext threadingContext,
+        Workspace workspace,
+        ProjectId projectId,
+        IAnalyzersCommandHandler commandHandler,
+        IDiagnosticAnalyzerService diagnosticAnalyzerService,
+        IAsynchronousOperationListenerProvider listenerProvider)
     {
-        private static readonly DiagnosticDescriptorComparer s_comparer = new DiagnosticDescriptorComparer();
+        _threadingContext = threadingContext;
+        Workspace = workspace;
+        ProjectId = projectId;
+        CommandHandler = commandHandler;
+        _diagnosticAnalyzerService = diagnosticAnalyzerService;
+
+        _listener = listenerProvider.GetListener(FeatureAttribute.SourceGenerators);
+        _workQueue = new AsyncBatchingWorkQueue(
+            DelayTimeSpan.Idle,
+            ProcessQueueAsync,
+            _listener,
+            _cancellationTokenSource.Token);
+    }
 
-        private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+    protected AnalyzerReference? AnalyzerReference
+    {
+        get => _analyzerReference_DoNotAccessDirectly;
+        set
+        {
+            Contract.ThrowIfTrue(_analyzerReference_DoNotAccessDirectly != null);
+            if (value is null)
+                return;
 
-        private BulkObservableCollection<BaseItem>? _items;
-        private ReportDiagnostic _generalDiagnosticOption;
-        private ImmutableDictionary<string, ReportDiagnostic>? _specificDiagnosticOptions;
-        private AnalyzerConfigData? _analyzerConfigOptions;
+            _analyzerReference_DoNotAccessDirectly = value;
 
-        public BaseDiagnosticAndGeneratorItemSource(Workspace workspace, ProjectId projectId, IAnalyzersCommandHandler commandHandler, IDiagnosticAnalyzerService diagnosticAnalyzerService)
-        {
-            Workspace = workspace;
-            ProjectId = projectId;
-            CommandHandler = commandHandler;
-            _diagnosticAnalyzerService = diagnosticAnalyzerService;
+            // Listen for changes that would affect the set of analyzers/generators in this reference, and kick off work
+            // to now get the items for this source.
+            Workspace.WorkspaceChanged += OnWorkspaceChanged;
+            _workQueue.AddWork();
         }
+    }
 
-        public Workspace Workspace { get; }
-        public ProjectId ProjectId { get; }
-        protected IAnalyzersCommandHandler CommandHandler { get; }
+    public abstract object SourceItem { get; }
 
-        public abstract AnalyzerReference? AnalyzerReference { get; }
+    [MemberNotNullWhen(true, nameof(AnalyzerReference))]
+    // Defer actual determination and computation of the items until later.
+    public bool HasItems => !_cancellationTokenSource.IsCancellationRequested;
```

Comment:
it is  fine for 'HasItems' to return true, even if hte items are computed asynchronously.  We do this in other places (like SourceGeneratedFileItemSource and AnalyzersFolderItemSource).

Reply:
Is the MemberNotNullWhen attribute still accurate?

Reply:
no.  let me fixup.